src/Access/SettingsConstraints.cpp

```diff
@@ -98,24 +101,20 @@ void SettingsConstraints::check(const Settings & current_settings, const Setting
             check(current_settings, value);
         }
 
+        SettingConstraintWritability new_value = SettingConstraintWritability::WRITABLE;
+        SettingConstraintWritability old_value = SettingConstraintWritability::WRITABLE;
+
         if (element.writability)
-        {
-            SettingConstraintWritability new_value = *element.writability;
-            SettingConstraintWritability old_value = SettingConstraintWritability::WRITABLE;
+            new_value = *element.writability;
 
-            auto it = constraints.find(element.setting_name);
-            if (it != constraints.end())
-                old_value = it->second.writability;
+        auto it = constraints.find(element.setting_name);
+        if (it != constraints.end())
+            old_value = it->second.writability;
 
-            if (new_value != old_value)
-            {
-                if (old_value == SettingConstraintWritability::CONST)
-                    throw Exception("Setting " + element.setting_name + " should not be changed", ErrorCodes::SETTING_CONSTRAINT_VIOLATION);
-
-                if (new_value == SettingConstraintWritability::CHANGEABLE_IN_READONLY &&
-                    old_value == SettingConstraintWritability::WRITABLE)
-                    throw Exception("Setting " + element.setting_name + " should not be changed", ErrorCodes::SETTING_CONSTRAINT_VIOLATION);
```

Comment:
Denying WRITEABLE -> CHANGABLE_IN_READONLY looked quite right to me (an extension for readonly users), why is this now removed? Maybe we should deny it at least for readonly users?

Reply:
Do you mean for the users with const/readonly constraints on the setting we're checking? If so, we seem to already do it.

The main reason is that WRITABLE is the default level, and it is WRITABLE regardless of if there was a constraint at all. So with it, it was actually not possible to set CHANGABLE_IN_READONLY at all.

Reply:
CREATE/ALTER queries are unavailable in `readonly>0`, it is true